File: Python/timer.py
import pygame
import time

# =================================================================================================
# Timer functions: Measuring time, delaying program execution, etc.

class CTimer:
    def __init__ (self, duration = 0):
        self.time = 0
        self.lapTime = 0
        self.duration = duration
        self.slack = 0


    def Start (self):
        self.time = pygame.time.get_ticks ()
        return self.time


    def Lap (self):
        self.lapTime = pygame.time.get_ticks () - self.time
        return self.lapTime


    def HasPassed (self, time = 0, restart = False):
        self.Lap ()
        if (time == 0):
            time = self.duration
        if (self.time > 0) and (self.lapTime < time):
            return False
        if (restart):
            self.Start ()
        return True


    def Delay (self):
            t = self.duration - self.slack - self.Lap ()
            if (t > 0):
                # pygame.time.wait (t)
                time.sleep (t / 1000.0)
            self.slack = self.Lap () - self.duration

# =================================================================================================

File: CSharp/timer.cs
using System.Threading;
using SDL2;

// =================================================================================================
// Timer functions: Measuring time, delaying program execution, etc.

public class Timer
{
    public int m_time;
    public int m_lapTime;
    public int m_duration;
    public int m_slack;

    public Timer(int duration = 0) { }


    public int Start()
    {
        return m_time = (int) SDL.SDL_GetTicks();
    }



    public int Lap()
    {
        return m_lapTime = (int) SDL.SDL_GetTicks() - m_time;
    }


    public bool HasPassed(int time = 0, bool restart = false)
    {
        Lap();
        if (time == 0)
            time = m_duration;
        if ((m_time > 0) && (m_lapTime < time))
            return false;
        if (restart)
            Start();
        return true;
    }


    public void Delay()
    {
        int t = m_duration - m_slack - Lap();
        if (t > 0)
            Thread.Sleep(t);
        m_slack = Lap() - m_duration;
    }

}

// =================================================================================================
